### PR TITLE
golint fixes for daemon/ package

### PR DESCRIPTION
**Pull request by MHBauer**
~~One linter advice yet:
`daemon/inspect.go:100:56: exported method ContainerExecInspect returns unexported type *daemon.execConfig, which can be annoying to use`
Suggestions for addressing it?~~

There are a surprising (to me) amount of things that became non-exported.

There some comments with "perhaps" in them that  I was not sure if I should take action on.

There are some renames with *Locking because all the functions did was add locking around an existing unexported function.

Some of the comments need work as to the exact details.

A partial history of individual commits can be found in my forked repo in the "split-commits" branch.

for #14756

### CODE REVIEW

**Comment by MHBauer** [comment]
I'm going to temporarily turn the linter off until it's agreed on how to implement the daemon/inspect.go advice.

**Comment by MHBauer** [comment]
fix for windows. kind of concerned that the cross compile did not fail. It should have built everything for Windows, right?

**Comment by aaronlehmann** [comment]
I'd suggest exporting execConfig. Any type returned by an exported function should also be exported.

File: daemon/container.go [hunk 1]
```diff
@@ -34,13 +34,13 @@ import (
 )
 
 var (
-	ErrNotATTY                 = errors.New("The PTY is not a file")
-	ErrNoTTY                   = errors.New("No PTY found")
-	ErrContainerStart          = errors.New("The container failed to start. Unknown error")
-	ErrContainerStartTimeout   = errors.New("The container failed to start due to timed out.")
+	// ErrContainerRootfsReadonly is returned when a container
+	// rootfs is marked readonly. (perhaps move to archive.go and
+	// make private, the only site of usage)
```
Comment (aaronlehmann) [inline daemon/container.go]: While we're at it, should this change to ErrContainerRootFSReadOnly? Or maybe something a little shorter?

File: daemon/container.go [hunk 2]
```diff
@@ -353,9 +359,14 @@ func (container *Container) cleanup() {
...
-	container.UnmountVolumes(false)
+	container.unmountVolumes(false)
 }
 
+// KillSig sends the container the given signal. This wrapper for the
+// host specific kill command prepares the container before attempting
+// to send the signal. An error is returned if the container is paused
+// or not running, or if there is a problem returned from the
+// underlying kill command.
 func (container *Container) KillSig(sig int) error {
```
Comment (aaronlehmann) [inline daemon/container.go]: Maybe this should be renamed, since KillSig implies it sends a kill signal instead of sending any signal.

File: daemon/container.go [hunk 3]
```diff
@@ -500,10 +517,11 @@ func (container *Container) Stop(seconds int) error {
 		}
 	}
 
-	container.LogEvent("stop")
+	container.logEvent("stop")
 	return nil
 }
 
+// Restart attempts to gracefully stop and then start the container.
```
Comment (aaronlehmann) [inline daemon/container.go]: Document the seconds argument?

File: daemon/container.go [hunk 4]
```diff
@@ -856,13 +866,15 @@ func (container *Container) monitorExec(execConfig *execConfig, callback execdri
 	return err
 }
 
-func (c *Container) Attach(stdin io.ReadCloser, stdout io.Writer, stderr io.Writer) chan error {
-	return attach(&c.StreamConfig, c.Config.OpenStdin, c.Config.StdinOnce, c.Config.Tty, stdin, stdout, stderr)
+// Attach connects to the containers TTY, delegating to standard
```
Comment (aaronlehmann) [inline daemon/container.go]: Small nit: "container's"

File: daemon/container_unix.go [hunk 5]
```diff
@@ -37,8 +37,11 @@ import (
 	"github.com/opencontainers/runc/libcontainer/label"
 )
 
+// DefaultPathEnv is unix style list of directories to search for executables.
```
Comment (aaronlehmann) [inline daemon/container_unix.go]: I'd suggest clarifying this by explicitly saying that the list is separated by : characters.

File: daemon/container_windows.go [hunk 6]
```diff
@@ -193,6 +185,7 @@ func (container *Container) PrepareStorage() error {
 	return nil
 }
 
+// CleanupStorage unprepares the layer after shutdown? FIXME
```
Comment (aaronlehmann) [inline daemon/container_windows.go]: Please address the FIXME.

File: daemon/daemon.go [hunk 7]
```diff
@@ -506,14 +516,17 @@ func (daemon *Daemon) GetByName(name string) (*Container, error) {
 	return e, nil
 }
 
-func (daemon *Daemon) Children(name string) (map[string]*Container, error) {
+// Children returns all child containers of the given name. The
+// containers are returned as a map from the container name to a
+// pointer to Container.
```
Comment (aaronlehmann) [inline daemon/daemon.go]: Should "given name" be "container with the given name"?

The second sentence could say "The containers are returned as a map indexed by container names".

Comment (aaronlehmann) [inline daemon/daemon.go]: Also, should lowercase "Children" now that the function is unexported.

File: daemon/daemon.go [hunk 8]
```diff
@@ -528,24 +541,28 @@ func (daemon *Daemon) Children(name string) (map[string]*Container, error) {
 	return children, nil
 }
 
-func (daemon *Daemon) Parents(name string) ([]string, error) {
+// Parents returns the names of the parents containers of the given
+// named container.
```
Comment (aaronlehmann) [inline daemon/daemon.go]: "Parents" should be lowercase.

Small nit: "names of the parent containers".

File: daemon/daemon.go [hunk 9]
```diff
@@ -743,12 +760,14 @@ func NewDaemon(config *Config, registryService *registry.Service) (daemon *Daemo
...
 
+// Shutdown stops the daemon.
 func (daemon *Daemon) Shutdown() error {
 	if daemon.containers != nil {
 		group := sync.WaitGroup{}
 		logrus.Debug("starting clean shutdown of all containers...")
 		for _, container := range daemon.List() {
-			c := container
+			var c *Container
+			c = container
```
Comment (aaronlehmann) [inline daemon/daemon.go]: Why this change?

File: daemon/daemon.go [hunk 10]
```diff
@@ -850,34 +873,40 @@ func (daemon *Daemon) Graph() *graph.Graph {
...
 	return daemon.execDriver
 }
 
-func (daemon *Daemon) ContainerGraph() *graphdb.Database {
-	return daemon.containerGraph
+func (daemon *Daemon) containerGraph() *graphdb.Database {
+	return daemon.containerGraphDB
 }
 
+// ImageGetCached return an image created before the given image with the same configuration.
```
Comment (aaronlehmann) [inline daemon/daemon.go]: s/return/returns/

What does it mean for an image to be created before another image? I feel like this could be clarified.

File: daemon/daemon_unix.go [hunk 11]
```diff
@@ -31,11 +31,15 @@ import (
 	"github.com/opencontainers/runc/libcontainer/label"
 )
 
+// Changes returns the list of changes between the container and it's
+// parent.
```
Comment (aaronlehmann) [inline daemon/daemon_unix.go]: s/it's/its/

File: daemon/daemon_unix.go [hunk 12]
```diff
@@ -31,11 +31,15 @@ import (
...
 func (daemon *Daemon) Changes(container *Container) ([]archive.Change, error) {
 	initID := fmt.Sprintf("%s-init", container.ID)
 	return daemon.driver.Changes(container.ID, initID)
 }
 
+// Diff returns the changes between the container and it's parent as
+// an archive.
```
Comment (aaronlehmann) [inline daemon/daemon_unix.go]: s/it's/its/

File: daemon/daemon_unix.go [hunk 13]
```diff
@@ -477,11 +481,13 @@ func setupInitLayer(initLayer string) error {
 	return nil
 }
 
-func (daemon *Daemon) NetworkApiRouter() func(w http.ResponseWriter, req *http.Request) {
+// NetworkAPIRouter perhaps we inline this into server-experimental
```
Comment (aaronlehmann) [inline daemon/daemon_unix.go]: Perhaps state this in a full sentence.

File: daemon/daemon_windows.go [hunk 14]
```diff
@@ -15,12 +15,18 @@ import (
 	"github.com/microsoft/hcsshim"
 )
 
-const DefaultVirtualSwitch = "Virtual Switch"
+const defaultVirtualSwitch = "Virtual Switch"
 
+// Changes returns the list of changes between the container and it's
+// parent.
+// CHECK
```
Comment (aaronlehmann) [inline daemon/daemon_windows.go]: s/it's/its/

Please address the CHECK comment before merging.

File: daemon/daemon_windows.go [hunk 15]
```diff
@@ -15,12 +15,18 @@ import (
...
 func (daemon *Daemon) Changes(container *Container) ([]archive.Change, error) {
 	return daemon.driver.Changes(container.ID, container.ImageID)
 }
 
+// Diff returns the changes between the container and it's parent as
+// an archive.
+// CHECK
```
Comment (aaronlehmann) [inline daemon/daemon_windows.go]: s/it's/its/

Please address the CHECK comment before merging.

File: daemon/daemon_unix.go [hunk 16]
```diff
@@ -477,11 +481,13 @@ func setupInitLayer(initLayer string) error {
...
+func (daemon *Daemon) NetworkAPIRouter() func(w http.ResponseWriter, req *http.Request) {
 	return nwapi.NewHTTPHandler(daemon.netController)
 }
 
-func (daemon *Daemon) RegisterLinks(container *Container, hostConfig *runconfig.HostConfig) error {
+// RegisterLinks writes the links to a file.
```
Comment (aaronlehmann) [inline daemon/daemon_unix.go]: s/RegisterLinks/registerLinks/

File: daemon/daemon_windows.go [hunk 17]
```diff
@@ -135,12 +141,14 @@ func isBridgeNetworkDisabled(config *Config) bool {
...
 	// Set the name of the virtual switch if not specified by -b on daemon start
 	if config.Bridge.VirtualSwitchName == "" {
-		config.Bridge.VirtualSwitchName = DefaultVirtualSwitch
+		config.Bridge.VirtualSwitchName = defaultVirtualSwitch
 	}
 	return nil, nil
 }
 
-func (daemon *Daemon) RegisterLinks(container *Container, hostConfig *runconfig.HostConfig) error {
+// RegisterLinks sets up links between containers and writes the
```
Comment (aaronlehmann) [inline daemon/daemon_windows.go]: s/RegisterLinks/registerlinks/

File: daemon/history.go [hunk 18]
```diff
@@ -22,10 +22,13 @@ func (history *History) Swap(i, j int) {
...
 }
 
+// Add the given container to history.
 func (history *History) Add(container *Container) {
 	*history = append(*history, container)
 }
 
-func (history *History) Sort() {
+// History implements the sort interface, can we inline calls to sort
+// rather than have this function?
```
Comment (aaronlehmann) [inline daemon/history.go]: IMHO, feel free to inline the calls.

File: daemon/inspect.go [hunk 19]
```diff
@@ -7,6 +7,9 @@ import (
 	"github.com/docker/docker/api/types"
 )
 
+// ContainerInspect returns lowlevel information about a
```
Comment (aaronlehmann) [inline daemon/inspect.go]: s/lowlevel/low-level/

File: daemon/inspect.go [hunk 20]
```diff
@@ -86,6 +89,14 @@ func (daemon *Daemon) getInspectData(container *Container) (*types.ContainerJSON
...
 }
 
+// ContainerExecInspect returns low-level information about the exec
+// command. An error is returned if the exec cannot be found.
+//
+// golnit complains exported method ContainerExecInspect returns
+// unexported type *daemon.execConfig, which can be annoying to use
+//
+// not sure what to do about the return value, perhaps we should
+// export execConfig
```
Comment (aaronlehmann) [inline daemon/inspect.go]: Pleaes export execConfig. Since it's returned by an exported function, the type should also be exported.

File: daemon/list.go [hunk 21]
```diff
@@ -17,6 +17,8 @@ func (daemon *Daemon) List() []*Container {
 	return daemon.containers.List()
 }
 
+// ContainersConfig is a struct for configuring the command to list
+// containers.
```
Comment (aaronlehmann) [inline daemon/list.go]: It would be useful to also document the struct members.

File: daemon/logs.go [hunk 22]
```diff
@@ -20,6 +22,8 @@ type ContainerLogsConfig struct {
 	Stop                 <-chan bool
 }
 
+// ContainerLogs hooks up a containers stdout and stderr streams
```
Comment (aaronlehmann) [inline daemon/logs.go]: s/containers/container's/

File: daemon/state.go [hunk 23]
```diff
@@ -111,10 +115,10 @@ func wait(waitChan <-chan struct{}, timeout time.Duration) error {
 	}
 }
 
-// WaitRunning waits until state is running. If state already running it returns
-// immediately. If you want wait forever you must supply negative timeout.
-// Returns pid, that was passed to SetRunning
-func (s *State) WaitRunning(timeout time.Duration) (int, error) {
+// WaitRunning waits until state is running. If state already running
```
Comment (aaronlehmann) [inline daemon/state.go]: s/state already/state is already/

Comment (aaronlehmann) [inline daemon/state.go]: s/WaitRunning/waitRunning/

File: daemon/state.go [hunk 24]
```diff
@@ -154,6 +159,7 @@ func (s *State) IsRunning() bool {
 	return res
 }
 
+// GetPid holds the process id of a container (perhaps this should be moved up into CommonContainer
```
Comment (aaronlehmann) [inline daemon/state.go]: Change to GetPID?

File: daemon/stats.go [hunk 25]
```diff
@@ -10,14 +10,18 @@ import (
 	"github.com/opencontainers/runc/libcontainer"
 )
 
+// ContainerStatsConfig holds information for configuring the runtime
+// behavior of a daemon.ContainerStats() call.
```
Comment (aaronlehmann) [inline daemon/stats.go]: Maybe this struct could be removed. The three items could just be arguments to ContainerStats.

File: daemon/stats_collector_unix.go [hunk 26]
```diff
@@ -122,9 +122,9 @@ func (s *statsCollector) run() {
 
 const nanoSeconds = 1e9
 
-// getSystemCpuUSage returns the host system's cpu usage in nanoseconds
+// getSystemCPUUSage returns the host system's cpu usage in nanoseconds
 // for the system to match the cgroup readings are returned in the same format.
```
Comment (aaronlehmann) [inline daemon/stats_collector_unix.go]: s/getSystemCPUUSage/getSystemCPUUsage/

Also, this sentence needs some work.

**Comment by aaronlehmann** [comment]
This is great. The new documentation comments are a pleasure to read.

There's a lot of stuff that got unexported. I'm not familiar enough with this code to weigh in on that. There may be certain functions that should be exported even though they're not currently used outside the package, to allow future flexibility. Hopefully people who understand this code well will comment on that.

File: daemon/container.go (hunk 1 above)
Comment (MHBauer) [inline daemon/container.go]: Does that mean you agree on moving it?

For a rename, ErrRootFSReadOnly? I cannot think of a way to make that much shorter, and I do not think "Container" adds anything informative. Errors like this can also be replaced simply with fmt.errorf as well, because they have no internal variables, only a non-specific message.

Comment (aaronlehmann) [inline daemon/container.go]: I think it's good practice to export the errors that get returned so package users can check for them, so I'd lean towards leaving it exported. Moving it to the file where it's used sounds okay but I don't think it matters much.

ErrRootFSReadOnly sounds good.

File: daemon/container_windows.go (hunk 6 above)
Comment (MHBauer) [inline daemon/container_windows.go]: @jhowardmsft May I ask for a comment on this method, and perhaps a review of the rest of this file as well?

File: daemon/daemon.go (hunk 7 above)
Comment (MHBauer) [inline daemon/daemon.go]: yes thanks. 

Now that I look at it again, I am not sure if the second sentence says anything more interesting than repeating the return signature.

File: daemon/daemon.go (hunk 8 above)
Comment (MHBauer) [inline daemon/daemon.go]: yup, I documented a lot of these before seeing how many could be unexported.

File: daemon/daemon.go (hunk 9 above)
Comment (MHBauer) [inline daemon/daemon.go]: whoops, relic of my understanding embedding. Definitely putting that back. Thanks.

File: daemon/daemon.go (hunk 10 above)
Comment (MHBauer) [inline daemon/daemon.go]: Reads like word salad to me. Rewrite.

Comment (MHBauer) [inline daemon/daemon.go]: // ImageGetCached returns the earliest created image that is a child
// of the image with imgID, that had the same config when it was
// created. nil is returned if a child cannot be found. An error is
// returned if the parent image cannot be found.

File: daemon/state.go (hunk 24 above)
Comment (MHBauer) [inline daemon/state.go]: Not indicated by golint, but it is in line with other changes that golint does suggest.

File: daemon/container_windows.go (hunk 6 above)
Comment (jhowardmsft) [inline daemon/container_windows.go]: prepareStorage is an extra step required on the Windows platform after mounting to ensure the storage is ready for use.

cleanupStorage is an extra step required on the Windows platform to ensure the storage is cleaned up before the storage is ready for use.

Note that these are being removed shortly to tidy up the Windows graph driver.

File: daemon/container_windows.go [hunk 27]
```diff
@@ -18,6 +18,8 @@ import (
 // the container. Docker has no context of what the default path should be.
```
Comment (jhowardmsft) [inline daemon/container_windows.go]: I'm a little surprised golint doesn't pickup that the godoc doesn't start "DefaultPathEnv...."

File: daemon/inspect.go [hunk 28]
```diff
@@ -86,6 +89,14 @@ func (daemon *Daemon) getInspectData(container *Container) (*types.ContainerJSON
 	return contJSONBase, nil
 }
 
+// ContainerExecInspect returns low-level information about the exec
+// command. An error is returned if the exec cannot be found.
+//
+// golnit complains exported method ContainerExecInspect returns
```
Comment (jhowardmsft) [inline daemon/inspect.go]: golint :)

File: daemon/state.go [hunk 29]
```diff
@@ -9,10 +9,13 @@ import (
...
 
+// State holds the current container state, and has methods to get and
+// set the state. Container has an embed, which allows all of the
+// functions defined against State to run against Container.
 type State struct {
 	sync.Mutex
-	Running           bool
-	Paused            bool
+	Running           bool // Why do we have both of these if a container
+	Paused            bool // cannot be paused and running at the same time?
```
Comment (jhowardmsft) [inline daemon/state.go]: Good point. I would put a FIXME here for someone to pickup later.

File: daemon/stop.go [hunk 30]
```diff
@@ -2,6 +2,12 @@ package daemon
 
 import "fmt"
 
+// ContainerStop looks for the given container and terminates it,
+// waiting the given number of second before forcefully killing the
```
Comment (jhowardmsft) [inline daemon/stop.go]: seconds

File: daemon/stop.go [hunk 31]
```diff
@@ -2,6 +2,12 @@ package daemon
...
+// container. If a negative number of second is given, ContainerStop
```
Comment (jhowardmsft) [inline daemon/stop.go]: seconds

File: docker/daemon.go [hunk 32]
```diff
@@ -266,6 +266,13 @@ func (cli *DaemonCli) CmdDaemon(args ...string) error {
 
 	logrus.Info("Daemon has completed initialization")
 
+	// FIXME: perhaps move the following log statement to the end
+	// of daemon.NewDaemon. Why? The exports ExecutionDriver() and
+	// GraphDriver() are only used here outside of the daemon
+	// package. It seems silly to me to export these two just to
```
Comment (jhowardmsft) [inline docker/daemon.go]: The last part of this is written in the context of the golint PR. When someone comes to FIXME, the context isn't obvious. I would be direct and say what needs changing.

File: daemon/container.go (hunk 1 above)
Comment (MHBauer) [inline daemon/container.go]: I think I understand why it does not matter where it is defined. I am still getting used to golang.

File: daemon/container.go (hunk 2 above)
Comment (MHBauer) [inline daemon/container.go]: It can become `Signal` in that the `kill` command is used to send signals. Or it can become `Kill` and `Kill` can become `KillForcefully` or some such that indicates it's unceremonious nature.

I would want a second and third opinion before making this change.

Comment (jhowardmsft) [inline daemon/container.go]: Actually, I'd prefer it doesn't become signal. Reason being, Windows really doesn't support signals in the same way that Linux does, and for containers we're not sending a signal anyway deep down in the implementation. Kill on it's own is more generic

File: daemon/container_windows.go (hunk 27 above)
Comment (MHBauer) [inline daemon/container_windows.go]: I think it only triggers on functions/methods and structs. I will format this comment so it matches.

**Comment by aaronlehmann** [comment]
Please rebase.

File: daemon/list.go (hunk 21 above)
Comment (MHBauer) [inline daemon/list.go]: documented individually now.

**Comment by MHBauer** [comment]
rebased.

**Comment by vdemeester** [comment]
@MHBauer needs another rebase, sorry :blush: :sweat_smile:.

**Comment by MHBauer** [comment]
Any other comments on KillSig?

I took out the comment on daemon to file a separate issue so there is discussion.

File: daemon/container.go (hunk 2 above)
Comment (vdemeester) [inline daemon/container.go]: @MHBauer it's `Kill` on `daemon` so I think it's safe to rename it `Kill` here too.

File: daemon/container.go [hunk 33]
```diff
@@ -220,18 +221,18 @@ func (container *Container) GetResourcePath(path string) (string, error) {
...
-func (container *Container) GetRootResourcePath(path string) (string, error) {
+func (container *Container) getRootResourcePath(path string) (string, error) {
 	// IMPORTANT - These are paths on the OS where the daemon is running, hence
 	// any filepath operations must be done in an OS agnostic way.
 	cleanPath := filepath.Join(string(os.PathSeparator), path)
 	return symlink.FollowSymlinkInScope(filepath.Join(container.root, cleanPath), container.root)
 }
 
-func (container *Container) ExportRw() (archive.Archive, error) {
+func (container *Container) exportContainerRw() (archive.Archive, error) {
```
Comment (vdemeester) [inline daemon/container.go]: I'm probably missing something here but : Why is this one rename ?

**Comment by vdemeester** [comment]
Hum, windows build is failing, `setRestarting` is re-defined somehow (`state.go:217`).

File: daemon/container.go (hunk 2 above)
Comment (MHBauer) [inline daemon/container.go]: Kill is already an exported function and is a wrapper around KillSig. KillSig does not have to be exported, and I will probably make that change, but I would like to keep the documentation if it is helpful.

File: daemon/container.go (hunk 33 above)
Comment (MHBauer) [inline daemon/container.go]: There is an already existing function in container_windows.go which causes a conflict if I downcase it.

**Comment by MHBauer** [comment]
Yea, bad rebase on my part. Fixing.

**Comment by LK4D4** [comment]
@MHBauer would you mind to rebase. I promise to review!

**Comment by LK4D4** [comment]
LGTM
I think we can fix minor comments in other PRs, because this is really painful to rebase and it is indeed fixes all goling warnings.

File: daemon/stats.go (hunk 25 above)
Comment (calavera) [inline daemon/stats.go]: out of scope

**Comment by calavera** [comment]
LGTM, brb, rebasing all my PRs that are affected by me merging this! :tada:
